### PR TITLE
feat: dataset editing + new dataset convergence

### LINKED ISSUE #273: Manual Edit UI for Readme and Meta
The manual dataset uploading UI is the same layout as the dataset editor.  Since that work is already in, let's modify this work to use those components.

**Add meta and readme editing to manual dataset creation flow**
- [ ] Make readme and meta tabs in the manual update screen editable

**Create a new `/edit` route with the same layout as the manual upload route**
- [ ] When the user clicks 'edit' in the history view, let's navigate to `/:username/:datasetname/edit` and show the same UI that we show for manual dataset creation, but with different state for editing the active dataset

- [ ] Confirm that the full editing flow works
  - [ ] Go to existing dataset history view
  - [ ] Click "edit"
  - [ ] Navigate to `/:username/:datasetname/edit`
  - [ ] Make a change to the readme and/or metadata
  - [ ] Make a change to the Title in the dataset header, it should also be reflected in the meta tab
  - [ ] Commit
  - [ ] Navigate back to `/history` showing the current version of the dataset

### PR DESCRIPTION
**Pull request by boandriy**
added functionality to edit Meta and Readme of a dataset

closes #273

### CODE REVIEW

**Comment by chriswhong** [comment]
Added a commit with a major refactor, consolidating the previous work done on this branch to edit readme and meta for an existing dataset with the new dataset creation flow that was recently merged into master.

Both new dataset creation and existing dataset are now consolidated into the same feature in `/src/datasetEditor`.  There are two top-level components, `NewDatasetEditor` and `ExistingDatasetEditor`, both of which render `DatasetEditorLayout` in different ways for their respective UXs.  

Both also share the same state, which includes a `Dataset` which will be mutated when the user edits components, loading and error booleans, and a `File` for body updates

```
export interface DatasetEditorState {
  dataset: Dataset
  loading: boolean
  error: string
  // file is used to store the file reference when POSTing a body via form data
  file?: File
}
```

There is a lot more to do, but as of dc08ea8 the user can create a new dataset from a CSV, and modify the readme and meta before committing.  They can also edit existing datasets via a button in the history list.

Still TODO:

*PUNT*
- [ ] We should only show the commit bar when there are changes.  This involves comparing the workingDataset with the original.  We can punt on this if it's too difficult, but it would be a nice effect to have the commit bar pop up as soon as the first edit happens.  (status quo: in `NewDatasetEditor`, it appears when the file is added, which might be what we want as it forces the user to have a body.  In `ExistingDatasetEditor` the prop is hard coded to `true` to it always shows.

- [x] Dataset Title Editing Behavior - In `NewDatasetEditor`, the dataset name is stored in component state and used when making the api call to create the new dataset.  However, in `ExistingDatasetEditor` we aren't currently handling a change to the dataset name.  Changing the name via the same method we had enabled in the past would require a trip to the server, and I think this would make the user lose their changes unless we add some code to prevent this.  Should the user be able to change the name of a dataset they are editing?  I tend to think we should disable the name edit field, and make a name change be a bigger deal with its own modal or something.  We need the name editor for new datasets for obvious reasons, but I think it's less necessary when editing an existing dataset.

- [x] Error handling for the commit api call in `ExistingDatasetEditor`.  We show a loading spinner in the commit button, but don't handle the error or display it.

- [x] User should be able to update the body of an existing dataset by dragging in a new file.  I think this won't work with the `commit` api call we are doing at the end of the `ExistingDatasetEditor` flow, so we should check with @Arqu to see if it's possible ATM.

*PUNT*
- [ ] We have existing components for `MetaEditor` and `ReadmeEditor`, but we build the editors into `Meta` and `Readme`.  I like the pattern of keeping these separate and just calling a different component for editing.  If it's not too much work lets separate them again.

Follow-on issues
- Add a structure editor

File: package.json
Comment (ramfox) [inline package.json]: Personally think we should be strict about `no-unused-vars`, is there a particular case we are coming up against?

File: src/chrome/DatasetCommitInfo.tsx
Comment (ramfox) [inline src/chrome/DatasetCommitInfo.tsx:41]: dunno about this one, looks like a really valuable line of code you deleted here

File: src/features/datasetEditor/ExistingDatasetEditor.tsx
Comment (ramfox) [inline src/features/datasetEditor/ExistingDatasetEditor.tsx]: stray log

File: src/features/datasetEditor/DatasetEditorLayout.tsx
Comment (ramfox) [inline src/features/datasetEditor/DatasetEditorLayout.tsx]: if there is no `meta.title` this reads `undefined - New Qri Dataset` (showing up on an older-ish dataset I have)

maybe we want `${dataset.meta?.title || dataset.name} - New Qri Dataset`, just to be on the safe side?

Also, this language `New Qri Dataset` shows up for both the existing & new versions of the editor. Not sure if this is on purpose.

File: src/features/datasetEditor/state/datasetEditorActions.ts
Comment (ramfox) [inline src/features/datasetEditor/state/datasetEditorActions.ts]: either remove or add a comment about when we should be able to comment this back in

File: src/features/dsComponents/meta/MetaEditor.tsx
Comment (ramfox) [inline src/features/dsComponents/meta/MetaEditor.tsx:238]: this is a more general comment about the `DatasetEditor` & its children components (`TabbedDatasetEditor`, `ComponentEditor`, `MetaEditor` & `ReadmeEditor`), they don't seem to be used any longer. If they aren't being kept around for future use, we should remove them in this pr or a follow up!

File: src/features/commits/DatasetCommitList.tsx
Comment (ramfox) [inline src/features/commits/DatasetCommitList.tsx]: we should have some check here if the user is logged in and it is their dataset, only then do we should the `Edit` button

File: src/features/commits/DatasetCommitList.tsx
Comment (ramfox) [inline src/features/commits/DatasetCommitList.tsx]: If I'm translating correctly, `NewVersionButton` would be the `Edit` button we now have. I think we can remove this comment as it is no longer relevant.

File: src/features/dataset/CompletionCheckmark.tsx
Comment (ramfox) [inline src/features/dataset/CompletionCheckmark.tsx:8]: 2 things: 
1) Is this used anywhere?
2) we seem to have two components called `CompletionCheckmark`, one defined in `src/features/dataset` & one defined in `src/features/datasetEditor`

File: src/features/dataset/DatasetHeaderLayout.tsx
Comment (ramfox) [inline src/features/dataset/DatasetHeaderLayout.tsx]: I like this naming! Wouldn't be opposed to changing `editable` to `userOwnsDataset` in other places that are appropriate (not necessary for this PR, just making a note as we move forward)

File: src/features/dataset/DatasetNameCommitModal.tsx
Comment (ramfox) [inline src/features/dataset/DatasetNameCommitModal.tsx]: Is this modal for a future flow?

File: src/features/dataset/modal/EditDatasetTitleModal.tsx
Comment (ramfox) [inline src/features/dataset/modal/EditDatasetTitleModal.tsx]: I don't understand this change, either it should change back or could use a comment clarifying the intention!

File: src/features/dataset/state/datasetActions.ts
Comment (ramfox) [inline src/features/dataset/state/datasetActions.ts]: It looks like most of these may have been replaced by the action in `src/features/datasetEditor/state/datasetEditorActions`

Unless they are being saved for future purposes, we should remove them.

File: src/features/dataset/state/datasetState.ts
Comment (ramfox) [inline src/features/dataset/state/datasetState.ts]: I think once we get rid of the actions in the above file, we can probably get rid of these as well.

**Review by ramfox — CHANGES_REQUESTED** [review]
This is looking great! Very fun to use!!

I noticed 1 backend specific issue while playing around, documented here: https://github.com/qri-io/qri/issues/1997. If you try to save a new version of a dataset, and only change an existing list of keywords, it will erroneous return "saving failed: no changes", even though the frontend is sending over all of the correct information.

The only two "bugs" I've found in the UI:
1) All users can see the "edit" button for all datasets, even ones they do not own. I think this should be fixed before we merge.
2) The documented bug that has been punted surrounding the commit bar, which sounds totally fine to punt, since pushing the button while having no changes does not cause any issues

I've also requested some changes, mostly around removing code that isn't being used.

Otherwise, looks fab!!

File: src/features/datasetEditor/DatasetEditorLayout.tsx
Comment (chriswhong) [reply src/features/datasetEditor/DatasetEditorLayout.tsx]: I moved the `<Head>` tags up to the respective wrapping components, and added the appropriate title for each along with the OR you suggested.

File: src/features/dataset/DatasetNameCommitModal.tsx
Comment (chriswhong) [reply src/features/dataset/DatasetNameCommitModal.tsx]: This is actually that modal that we had in place to edit the title with a commit button, but we've decided to not use it for now.  The component is still around if we choose to use it again.

File: src/features/dataset/CompletionCheckmark.tsx
Comment (chriswhong) [reply src/features/dataset/CompletionCheckmark.tsx:8]: I believe these are for the "tic tacs" for showing which components have changed.  We aren't yet ready to use these but we can clean this up later

File: src/features/dataset/modal/EditDatasetTitleModal.tsx
Comment (chriswhong) [reply src/features/dataset/modal/EditDatasetTitleModal.tsx]: These were changed as the actions they use are no longer present and it was causing compiling errors.  I just removed the component as we are no longer using it.

File: src/features/dsComponents/meta/MetaEditor.tsx
Comment (chriswhong) [reply src/features/dsComponents/meta/MetaEditor.tsx:238]: There may be bits in here we want to re-use, I'll make an issue for cleanup of the old editor components

**Pull request merged**
